File: src/middle/abstract.jl
const recursionLimit = 10

_typeof(x) = error("invalid constant $x::$(typeof(x))")
_typeof(x::Union{Bits,Number,String,Tag,RMethod,Pack}) = x

# TODO constants should be converted early to empty tuples of constant type.
# If literals appear in the IR they are runtime values (eg args to foreign calls)
exprtype(ir, x) = IRTools.exprtype(ir isa Block ? ir.ir : ir, x, typeof = _typeof)
exprtype(ir, xs::Union{AbstractVector,Tuple}) = map(x -> exprtype(ir, x), xs)

function prepare_ir!(ir)
  ir = ir |> globals |> IRTools.expand!
  # TODO use the API
  for b in ir.blocks
    b.argtypes .= (⊥,)
    for i in 1:length(b.stmts)
      T = b.stmts[i][2].type
      T == Any && (T = ⊥)
      b.stmts[i] = (b.stmts[i][1], stmt(b.stmts[i][2], type = T))
    end
  end
  ir = looped(ir)
  return ir
end

function blockargs!(b, args)
  changed = false
  for i = 1:length(argtypes(b))
    issubset(args[i], argtypes(b)[i]) && continue
    argtypes(b)[i] = union(argtypes(b)[i], args[i])
    changed = true
  end
  return changed
end

struct Parent
  sig::Tuple
  depth::Int
end

struct Redirect
  to::Tuple
end

const Sig = Union{Tuple,Binding}

mutable struct GlobalFrame
  type
  deps::Set{Sig}
  edges::Set{Sig}
end

GlobalFrame(T) = GlobalFrame(T, Set(), Set())

mutable struct Frame
  parent::Parent
  ir::LoopIR
  deps::Set{Sig}
  edges::Set{Sig}
  rettype
end

Frame(P, ir::LoopIR, T = ⊥) =
  Frame(P, ir, Set(), Set(), T)

Base.show(io::IO, ::Frame) = print(io, "Frame(...)")

function frame(P, ir::IR, args...)
  ir = prepare_ir!(copy(ir))
  @assert length(arguments(ir.ir)) == length(args)
  argtypes(ir.body[1]) .= args
  return Frame(P, ir)
end

function const_frame(P, T, F, sig...)
  name = F isa Tag ? F : F.name
  ir = IR(meta = FuncInfo(name, trampoline = true))
  foreach(T -> argument!(ir, T), sig)
  r = push!(ir, stmt(Expr(:tuple), type = T))
  return!(ir, r)
  return Frame(P, looped(ir), T)
end

struct Inference
  defs::Definitions
  meths::EagerCache
  deps::IdDict{Any,Set{Caches.NFT}}
  frames::IdDict{Sig,Union{Frame,GlobalFrame,Redirect}}
  queue::WorkQueue{Tuple}
end

function Inference(defs::Definitions, meths::EagerCache)
  Inference(defs, meths, IdDict(), Dict(), WorkQueue{Tuple}())
end

function sig(inf::Inference, T)
  T == () && return ()
  fr = inf.frames[T]
  fr isa Redirect ? sig(inf, fr.to) : T
end

frame(inf::Inference, T) = inf.frames[sig(inf, T)]

parent(inf, T) = sig(inf, frame(inf, T).parent.sig)

function recursionDepth(inf, T, F)
  sigs = []
  while T != ()
    T[1] == F && return frame(inf, T).parent.depth+1
    T = parent(inf, T)
  end
  return 1
end

function frame!(inf::Inference, name::Binding)
  get!(inf.frames, name) do
    T, deps = Caches.trackdeps(() -> inf.defs[name])
    inf.deps[name] = deps
    if T isa Binding
      parent = frame!(inf, T)
      push!(parent.edges, name)
      T = parent.type
    end
    GlobalFrame(T)
  end
end

function irframe!(inf::Inference, P, ir, F, Ts...)
  inf.frames[(F, Ts...)] = frame(P, ir, Ts...)
  update!(inf, (F, Ts...))
  return frame(inf, (F, Ts...))
end

function frame!(inf::Inference, P, meth::RMethod, Ts...)
  if meth.partial
    meth.func(Ts...)
  elseif haskey(inf.frames, (meth, Ts...))
    frame(inf, (meth, Ts...))
  elseif P.depth > recursionLimit
    mergeFrames(inf, P.sig, (meth, Ts...))
  else
    meth.func isa IR || error("No IR for $meth: $Ts")
    irframe!(inf, P, meth.func, meth, Ts...)
  end
end

function frame!(inf::Inference, P, F, Ts)
  haskey(inf.frames, (F, Ts)) && return frame(inf, (F, Ts))
  inf.frames[(F, Ts)] = Frame(P, looped(IR()))
  update!(inf, (F, Ts))
  return frame(inf, (F, Ts))
end

# TODO some methods become unreachable, remove them somewhere?
function mergeFrames(inf::Inference, T, F)
  sigs = filter(t -> t[1] == F[1], [stack(inf, T).frames..., F])
  @assert length(sigs) > 1
  sig = reduce((a, b) -> union.(a, b), sigs)
  P = inf.frames[sigs[1]].parent.sig
  fr = frame!(inf, Parent(P, recursionLimit), sig...)
  for s in sigs
    s === sig && continue
    if haskey(inf.frames, s)
      if inf.frames[s] isa Redirect
        @assert sig == inf.frames[s].to # TODO figure out whether to move backedges here
      else
        union!(fr.edges, inf.frames[s].edges)
        push!(fr.edges, s) # propagate deletions
      end
    end
    inf.frames[s] = Redirect(sig)
  end
  return fr
end

function infercall!(inf::Inference, P, sig...)
  any(==(⊥), sig) && return ⊥
  P = Parent(P, recursionDepth(inf, P, sig[1]))
  fr = frame!(inf, P, sig...)
  fr isa Frame || return fr
  pf = inf.frames[P.sig]
  pf isa Redirect && return
  push!(pf.deps, sig)
  push!(fr.edges, P.sig)
  return fr.rettype
end

function infercall!(inf::Inference, P, ir, ex::Expr)
  infercall!(inf, P, exprtype(ir, ex.args)...)
end

function cleardeps!(inf::Inference, sig)
  fr = inf.frames[sig]
  for dep in fr.deps
    delete!(inf.frames[dep].edges, sig)
  end
  empty!(fr.deps)
  return
end

function update_dispatcher!(inf::Inference, sig)
  (ir, ret), deps = Caches.trackdeps(() -> dispatcher(inf, sig...))
  inf.deps[sig] = deps
  fr = inf.frames[sig]
  fr.ir = looped(IRTools.expand!(ir))
  if !issubset(ret, fr.rettype)
    fr.rettype = ret
    foreach(s -> push!(inf.queue, s), fr.edges)
  end
end

function update!(inf::Inference, sig)
  cleardeps!(inf, sig)
  sig[1] isa RMethod || return update_dispatcher!(inf, sig)
  fr = inf.frames[sig]
  ret = ⊥
  path = Path()
  reachable = Set([path])
  while path != nothing
    @label loop
    bl = block(fr.ir, path)
    for (v, st) in bl
      if isexpr(st, :call) && st.expr.args[1] isa WebAssembly.Instruction
        op = st.expr.args[1]
        Ts = exprtype(bl.ir, st.expr.args[2:end])
        if all(isvalue, Ts) && haskey(wasmPartials, op)
          T = wasmPartials[op](Ts...)
          bl.ir[v] = stmt(bl[v], type = T)
        end
      elseif isexpr(st, :call)
        T = infercall!(inf, sig, bl, st.expr)
        T == nothing && return ⊥
        T == ⊥ && break
        bl.ir[v] = stmt(bl[v], type = T)
      elseif isexpr(st, :pack)
        Ts = exprtype(bl.ir, st.expr.args)
        any(==(⊥), Ts) && break
        bl.ir[v] = stmt(bl[v], type = pack(Ts...))
      elseif isexpr(st, :global)
        g = frame!(inf, st.expr.args[1])
        push!(fr.deps, st.expr.args[1])
        push!(g.edges, sig)
        g.type == ⊥ && break
        bl.ir[v] = stmt(bl[v], type = g.type)
      elseif isexpr(st, :(=)) && st.expr.args[1] isa Binding
        T = exprtype(bl.ir, st.expr.args[2])
        T == ⊥ && break
        bl.ir[v] = stmt(st, type = T)
      elseif isexpr(st, :loop)
        l = loop(bl)
        blockargs!(l.body[1], argtypes(bl))
        path = Path([path.parts..., (1,1)])
        @goto loop
      elseif isexpr(st, :branch)
        br = st.expr
        if isreturn(br)
          ret = union(ret, exprtype(bl.ir, only(arguments(br))))
        elseif isunreachable(br)
          break
        else
          cond = exprtype(bl.ir, something(br.args[2], RBool(true)))
          @assert tag(cond) == tag"common.Bool"
          cond == RBool(false) && continue
          p, rr = nextpath(fr.ir, path, br.args[1])
          rr = rr || pin!(fr.ir, path, length(p))
          @assert !rr || p < path "unimplemented"
          args = exprtype(bl.ir, arguments(br))
          push!(reachable, p)
          if (blockargs!(block(fr.ir, p), args) || rr) && p < path
            path = p
            @goto loop
          end
          cond == RBool(true) && break
        end
      else
        error("Unknown expr type $(st.expr.head)")
      end
    end
    while true
      path = nextpath(fr.ir, path)
      (isnothing(path) || path in reachable) && break
    end
  end
  if !issubset(ret, fr.rettype)
    fr.rettype = ret
    foreach(s -> push!(inf.queue, s), fr.edges)
  end
  return
end

# TODO remove backedges, so we don't do redundant work
function Base.delete!(inf::Inference, sig::Union{Tuple,Binding})
  haskey(inf.frames, sig) || return
  fr = inf.frames[sig]
  # cleardeps!(inf, sig)
  delete!(inf.frames, sig)
  delete!(inf.deps, sig)
  fr isa Redirect && return
  foreach(loc -> delete!(inf, loc), fr.edges)
end

# Virtual stack traces

struct Stack
  frames::Vector{Any}
end

Stack() = Stack([])

function stack(inf::Inference, T)
  st = Stack()
  while haskey(inf.frames, T)
    pushfirst!(st.frames, T)
    T = parent(inf, T)
  end
  return st
end

function Base.show(io::IO, stack::Stack)
  println(io, "Abstract stack trace:")
  for (f, Ts...) in stack.frames
    print(io, f, ": ")
    println(io, f isa RMethod ? Ts : Ts[1])
  end
end

struct CompileError <: Exception
  error
  stack::Stack
end

function Base.showerror(io::IO, err::CompileError)
  println(io, "Compiler error at")
  println(io, err.stack)
  Base.showerror(io, err.error)
end

# Inference Loop

function infer!(inf::Inference; partial = false)
  while !isempty(inf.queue)
    sig = pop!(inf.queue)
    try
      update!(inf, sig)
    catch e
      partial && break
      rethrow(CompileError(e, stack(inf, sig)))
    end
  end
  return inf
end

# Results and caching

struct Inferred
  inf::Inference
  results::Caches.Dict{Any,Any}
end

function Inferred(defs::Definitions, meths::EagerCache)
  inf = Inference(defs, meths)
  return Inferred(inf, Caches.Dict())
end

Caches.iscached(i::Inferred, k) = iscached(i.results, k)

function Base.getindex(i::Inferred, sig)
  iscached(i, sig) && return i.results[sig]
  # Don't let inference dependencies leak
  _, deps = Caches.trackdeps() do
    frame!(i.inf, Parent((), 1), sig...)
    infer!(i.inf)
  end
  @assert isempty(deps)
  for (k, fr) in i.inf.frames
    (k isa Binding || iscached(i, k)) && continue
    i.results[k] = fr isa Redirect ? fr : (prune!(unloop(fr.ir)) => fr.rettype)
  end
  return i.results[sig]
end

Caches.fingerprint(i::Inferred) = fingerprint(i.results)

function Caches.reset!(i::Inferred; deps = [])
  print = fingerprint(deps)
  for (x, deps) in i.inf.deps
    Base.issubset(deps, print) || delete!(i.inf, x)
  end
  for k in keys(i.results)
    haskey(i.inf.frames, k) || delete!(i.results, k)
  end
end

File: src/middle/abstract.ts
import { unreachable, expand, Anno, Block, Expr, Branch, Pipe, expr, prune, asType } from '../utils/ir.js'
import { LoopIR, looped, Path, block, nextpath, nextpathTo, blockargs, loop, unloop } from './loop.js'
import { MatchMethods, dispatcher } from './patterns.js'
import {
  Type, repr, union, issubset as iss, isValue, pack, Closure,
  tag, tagOf, parts, String as RString, Ref, nil, Any, disjuncts
} from '../frontend/types.js'
import { wasmPartials } from '../backend/wasm.js'
import {
  MIR, IRValue, Binding, Dispatch, Method, Definitions, StringRef, JS, Invoke, Closure as XClosure,
  Global, SetGlobal, Wasm, callargs
} from '../frontend/modules.js'
import { Lowered } from '../frontend/lower.js'
import { Def } from '../dwarf/index.js'
import { WorkQueue } from '../utils/fixpoint.js'
import { hash, HashSet, only, some } from '../utils/map.js'
import { trackdeps, Map as CacheMap, fingerprint, Caching, withtime } from '../utils/cache.js'
import { isEqual } from '../utils/isEqual.js'
import { Instruction } from '../wasm/wasm.js'
import { Traced } from './tracer.js'
import { binding } from '../utils/options.js'
import { Lowering, Transform, partialPrimitive, transformPrimitive } from './prim_map.js'

const recursionLimit = 10

export { key, Sig, Inference, Inferred, inferexpr, infercall, issubset, maybe_union, traitType, withTraits }

function key(sig: Sig): string {
  const [f, ...Ts] = sig
  return `${f[hash]}:${Ts.map(repr).join(', ')}`
}

function maybe_union(x: Anno<Type>, y: Anno<Type>): Anno<Type> {
  if (x === unreachable) return y
  if (y === unreachable) return x
  return union(x, y)
}

type Func = Dispatch | Method
type Sig = [Dispatch, Type, Type] | [Method, ...Type[]]
type AIR = LoopIR<IRValue, Type>

function prepare_ir(ir: MIR, args: Type[]): AIR {
  const l = looped(expand(ir.clone()))
  if (l.ir.block(1).args.length !== args.length) throw new Error('argument length mismatch')
  const b = l.body[0].block(1)
  for (let i = 0; i < args.length; i++) b.bb.args[i][1] = args[i]
  return l
}

function striptypes(code: MIR): MIR {
  for (const [v, st] of code)
    if (!(st.expr instanceof Wasm)) code.setType(v, unreachable)
  for (const bl of code.blocks())
    for (const arg of bl.bb.args) arg[1] = unreachable
  return code
}

class Stack {
  constructor(readonly sigs: readonly Sig[] = []) { }
  push(sig: Sig): Stack { return new Stack([...this.sigs, sig]) }
  select(f: Func): Sig[] { return this.sigs.filter(([g]) => isEqual(g, f)) }
  toString(): string {
    return ['Abstract stack trace:', ...this.sigs.map(([f, ...Ts]) =>
      `${f.toString()}: ${f instanceof Method ? `(${Ts.map(repr).join(', ')})` : repr(Ts[0])}`)].join('\n')
  }
}

class GlobalFrame {
  constructor(
    public type: Anno<Type>,
    readonly deps = new Set<never>(),
    readonly edges = new Set<string>()) { }
  clone(): GlobalFrame { return new GlobalFrame(this.type, new Set(this.deps), new Set(this.edges)) }
}

class Frame {
  readonly key: string
  deps = new Set<string>()
  edges = new Set<string>()
  rettype: Anno<Type> = unreachable
  constructor(readonly sig: Sig, readonly stack: Stack, private _ir?: AIR) { this.key = key(sig) }
  get ir(): AIR { return some(this._ir, `Missing IR for inference frame ${this.key}`) }
  set ir(ir: AIR) { this._ir = ir }
  static create(stack: Stack, ir: MIR, f: Func, ...args: Type[]): Frame {
    return new Frame([f, ...args] as Sig, stack, prepare_ir(ir, args))
  }
  ghost(): Frame {
    const out = new Frame(this.sig, this.stack)
    out.deps = new Set(this.deps)
    out.edges = new Set(this.edges)
    out.rettype = this.rettype
    return out
  }
}

class Inference {
  deps = new Map<string, Set<bigint>>()
  frames = new Map<string, Frame>()
  redirects = new Map<string, Sig>()
  globals = new Map<string, GlobalFrame>()
  queue = new WorkQueue<string>()
  constructor(readonly defs: Definitions, readonly lowered: Lowered, readonly meths: MatchMethods,
    readonly traced?: Traced, readonly transforms?: Inferred) { }
  reuse(ch: this): this {
    this.deps = new Map(Array.from(ch.deps, ([k, deps]) => [k, new Set(deps)]))
    this.frames = new Map(Array.from(ch.frames, ([k, fr]) => [k, fr.ghost()]))
    this.redirects = new Map(ch.redirects)
    this.globals = new Map(Array.from(ch.globals, ([k, fr]) => [k, fr.clone()]))
    this.queue = new WorkQueue<string>()
    return this
  }
}

function resolve(inf: Inference, T: Sig): Sig {
  const to = inf.redirects.get(key(T))
  return to ? resolve(inf, to) : T
}

function lookup(inf: Inference, T: Sig): Frame | undefined {
  return inf.frames.get(key(resolve(inf, T)))
}

function globalFrame(inf: Inference, name: Binding): GlobalFrame {
  const key = name[hash]
  const existing = inf.globals.get(key)
  if (existing) return existing
  const [T, deps] = trackdeps(() => inf.defs.global(name))
  inf.deps.set(key, deps)
  let type = T
  if (type instanceof Binding) {
    const parentFrame = globalFrame(inf, type)
    parentFrame.edges.add(key)
    type = parentFrame.type
  }
  const frame = new GlobalFrame(type)
  inf.globals.set(key, frame)
  return frame
}

function call(inf: Inference, stack: Stack, sig: Sig): Frame {
  const T = resolve(inf, sig)
  const existing = inf.frames.get(key(T))
  if (existing) return existing
  if (T[0] instanceof Method && stack.select(T[0]).length >= recursionLimit)
    return widen(inf, stack, T)
  return frame(inf, stack, T)
}

function frame(inf: Inference, stack: Stack, sig: Sig): Frame {
  const existing = lookup(inf, sig)
  if (existing) return existing
  const [f, ...Ts] = sig
  const k = key(sig)
  const [result, deps] = trackdeps(() => {
    if (f instanceof Method && transformPrimitive(f))
      return inf.transforms?.get([f, ...Ts] as Sig)
    return inf.traced?.trace(f, ...Ts)
  })
  if (result) {
    const [ir, ret] = result
    const fr = Frame.create(stack, ir, f, ...Ts)
    fr.rettype = ret
    inf.deps.set(k, deps)
    inf.frames.set(k, fr)
    return fr
  }
  if (f instanceof Method && !transformPrimitive(f)) {
    const [ir, ideps] = trackdeps(() => inf.lowered.ir(f))
    for (const dep of ideps) deps.add(dep)
    inf.frames.set(k, Frame.create(stack, ir, f, ...Ts))
  } else {
    const tag = f instanceof Method ? f.name : f
    inf.frames.set(k, new Frame(sig, stack, looped(MIR(Def(tag.path)))))
  }
  inf.deps.set(k, deps)
  update(inf, k)
  return some(lookup(inf, sig))
}

function redirect(inf: Inference, from: Sig, to: Frame): void {
  const k = key(from)
  const fr = inf.frames.get(k)
  inf.redirects.set(k, to.sig)
  if (!fr) return
  cleardeps(inf, k)
  inf.frames.delete(k)
  for (const c of fr.edges) {
    const caller = inf.frames.get(c)
    if (!caller) continue
    caller.deps.delete(k)
    caller.deps.add(to.key)
    to.edges.add(c)
    inf.queue.push(c)
  }
}

function unionSig(a: Sig, b: Sig): Sig {
  const [f, ...as] = a
  const [, ...bs] = b
  return [f, ...as.map((x, i) => union(x, bs[i]))] as Sig
}

// TODO some methods become unreachable
function widen(inf: Inference, stack: Stack, F: Sig): Frame {
  const sigs = [...stack.select(F[0]).map(s => resolve(inf, s)), F]
  const widened = sigs.reduce(unionSig)
  const fr = frame(inf, stack, widened)
  for (const s of sigs) if (key(s) !== fr.key) redirect(inf, s, fr)
  return fr
}

function infercall(inf: Inference, P: Sig, F: Func, ...Ts: Anno<Type>[]): Anno<Type> | undefined {
  if (Ts.some(t => t === unreachable)) return unreachable
  if (F instanceof Method) {
    const partial = partialPrimitive(F)
    if (partial) return withTraits(T => traitResult(some(infercall(inf, P, ...traitSig(T)))),
      () => partial(...Ts as Type[]))
  }
  const stack = some(inf.frames.get(key(P))).stack.push(P)
  const fr = call(inf, stack, [F, ...Ts as Type[]] as Sig)
  const caller = inf.frames.get(key(P))
  if (!caller) return
  caller.deps.add(fr.key)
  fr.edges.add(caller.key)
  return fr.rettype
}

function inferexpr(inf: Inference, P: Sig, ir: MIR | Block<MIR>, ex: Expr<IRValue>): Anno<Type> | undefined {
  if (ex.head === 'call') {
    const calleeT = ir.type(ex.body[0])
    if (calleeT === unreachable) return unreachable
  }
  let [F, Ts] = callargs(ir, ex)
  return infercall(inf, P, F, ...Ts.map(x => ir.type(x)))
}

function cleardeps(inf: Inference, k: string): void {
  const fr = some(inf.frames.get(k))
  for (const dep of fr.deps) (inf.frames.get(dep) ?? inf.globals.get(dep))?.edges.delete(k)
  fr.deps.clear()
}

function issubset(x: Anno<Type>, y: Anno<Type>): boolean {
  if (x === unreachable) return true
  if (y === unreachable) return false
  return iss(x, y)
}

function settype(inf: Inference, fr: Frame, ret: Anno<Type>): void {
  if (issubset(ret, fr.rettype)) return
  fr.rettype = ret
  for (const s of fr.edges) inf.queue.push(s)
}

function framecode(inf: Inference, fr: Frame, sig: Sig): MIR {
  const callee = call(inf, fr.stack.push(fr.sig), sig)
  fr.deps.add(callee.key)
  callee.edges.add(fr.key)
  return prune(unloop(callee.ir))
}

function lowering(inf: Inference, fr: Frame): Lowering {
  const code: Lowering = {
    ir(f: Func, ...Ts: Type[]): MIR {
      if (!(f instanceof Method)) return striptypes(framecode(inf, fr, [f, ...Ts] as Sig))
      const transform = transformPrimitive(f)
      return transform ? transform(code, f, ...Ts) : inf.lowered.ir(f).clone()
    }
  }
  return code
}

function update_transform(inf: Inference, fr: Frame, f: Method, Ts: Type[], transform: Transform) {
  const [code, deps] = trackdeps(() => transform(lowering(inf, fr), f, ...Ts))
  inf.deps.set(fr.key, deps)
  fr.ir = prepare_ir(code, Ts)
}

function update_dispatcher(inf: Inference, fr: Frame, func: Dispatch, F: Type, Ts: Type) {
  const [[ir, ret], deps] = trackdeps(() => dispatcher(inf, func, F, Ts))
  inf.deps.set(fr.key, deps)
  fr.ir = looped(expand(ir))
  settype(inf, fr, ret)
}

function update(inf: Inference, k: string): void {
  const fr = inf.frames.get(k)
  if (!fr) return
  cleardeps(inf, k)
  const [f, ...Ts] = fr.sig
  if (!(f instanceof Method)) return update_dispatcher(inf, fr, f, Ts[0], Ts[1])
  const transform = transformPrimitive(f)
  if (transform) update_transform(inf, fr, f, Ts, transform)
  let ret: Anno<Type> = unreachable
  let path: Path | null = new Path()
  const reachable = new HashSet<Path>([path])
  update: while (path) {
    const bl = block(fr.ir, path)
    for (const [v, st] of bl) {
      const ex = st.expr
      if (ex instanceof Wasm) {
        if (ex.isImport()) continue
        const instr = ex.callee as Instruction
        const op = instr.kind === 'op' ? instr.name : ''
        const Ts = ex.body.map(x => bl.type(x))
        if (Ts.every(t => t !== unreachable) && Ts.every(t => isValue(t)) && wasmPartials.has(op)) {
          const T = some(wasmPartials.get(op))(...Ts)
          bl.ir.setType(v, T)
        }
      } else if (['call', 'invoke'].includes(ex.head)) {
        const T = inferexpr(inf, fr.sig, bl, ex)
        if (T === undefined) return
        if (T === unreachable) break
        bl.ir.setType(v, T)
      } else if (ex.head === 'pack') {
        const Ts = ex.body.map(x => bl.type(x))
        if (Ts.some(t => t === unreachable)) break
        bl.ir.setType(v, pack(...Ts as Type[]))
      } else if (ex instanceof XClosure) {
        const Ts = ex.args.map(x => bl.type(x))
        if (Ts.some(t => t === unreachable)) break
        bl.ir.setType(v, Closure(ex.method, ...Ts as Type[]))
      } else if (ex instanceof Global) {
        const g = globalFrame(inf, ex.binding)
        fr.deps.add(ex.binding[hash])
        g.edges.add(k)
        if (g.type === unreachable) break
        bl.ir.setType(v, g.type)
      } else if (ex instanceof SetGlobal) {
        if (bl.type(ex.value) === unreachable) break
        bl.ir.setType(v, nil)
      } else if (ex.head === 'loop') {
        const inner = some(loop(bl))
        blockargs(inner.body[0].block(1), bl.argtypes.map(t => asType(t)))
        path = new Path([...path.parts, [1, 1]])
        continue update
      } else if (st.expr instanceof Branch) {
        const br = st.expr
        if (br.isreturn()) {
          ret = maybe_union(ret, bl.type(br.args[0]))
        } else if (br.isunreachable()) {
          break
        } else {
          const condT = br.isconditional() ? asType(bl.type(br.when)) : Type(true)
          if (!tag('common.integer/Bool').isEqual(tagOf(condT))) throw new Error('branch condition must be Bool')
          if (isEqual(condT, Type(false))) continue
          let [p, rr] = nextpathTo(fr.ir, path, br.target)
          if (rr && !p.lt(path)) throw new Error('unimplemented')
          const args = br.args.map(a => asType(bl.type(a)))
          reachable.add(p)
          if ((blockargs(block(fr.ir, p), args) || rr) && p.lt(path)) {
            path = p
            continue update
          }
          if (isEqual(condT, Type(true))) break
        }
      } else if (ex.head === 'tuple') {
        if (!isValue(asType(st.type))) throw new Error('tuple without type')
      } else if (ex instanceof StringRef) {
        bl.ir.setType(v, RString())
      } else if (ex instanceof JS) {
        bl.ir.setType(v, Ref)
      } else throw new Error(`Unknown expr type ${ex.head}`)
    }
    while (true) {
      path = nextpath(fr.ir, path)
      if (!path || reachable.has(path)) break
    }
  }
  settype(inf, fr, ret)
}

// Invalidate a signature and everything that depends on it.
function remove(inf: Inference, k: string) {
  const fr = inf.frames.get(k) ?? inf.globals.get(k)
  if (fr instanceof Frame) cleardeps(inf, k)
  inf.frames.delete(k)
  inf.globals.delete(k)
  inf.redirects.delete(k)
  inf.deps.delete(k)
  for (const loc of fr?.edges ?? []) remove(inf, loc)
}

class CompileError extends Error {
  constructor(readonly error: unknown, readonly vstack: Stack) {
    super('CompileError')
    Object.setPrototypeOf(this, new.target.prototype)
  }
  toString(): string {
    return `Compiler error at\n${this.vstack.toString()}\n${String(this.error)}`
  }
}

// Inference Loop

function infer(inf: Inference, { partial = false }: { partial?: boolean } = {}): Inference {
  while (!inf.queue.empty) {
    const k = inf.queue.pop()
    try {
      update(inf, k)
    } catch (e) {
      if (partial) break
      const fr = inf.frames.get(k)
      if (!fr) throw e
      throw new CompileError(e, fr.stack.push(fr.sig))
    }
  }
  return inf
}

// Results and caching

function redirectCalls(inf: Inference, code: MIR): MIR {
  const pr = new Pipe(code)
  for (const [v, st] of pr) {
    if (!(st.expr instanceof Invoke)) continue
    if (partialPrimitive(st.expr.method)) continue
    const S = st.expr.body.map(x => asType(pr.type(x)))
    const from: Sig = [st.expr.method, ...S]
    const [_, ...T] = resolve(inf, from)
    if (isEqual(S, T)) continue
    pr.delete(v)
    const args = st.expr.body.map((x, i) =>
      isEqual(S[i], T[i]) ? x : pr.push(pr.stmt(expr('cast', x), { type: T[i], src: st.src })))
    const redirected = pr.push({ ...st, expr: new Invoke(st.expr.method, args) })
    pr.replace(v, redirected)
  }
  return pr.finish()
}

function forward(inf: Inference, sig: Sig): [MIR, Anno<Type>] {
  const [, ...S] = sig
  const [F, ...T] = resolve(inf, sig)
  if (!(F instanceof Method)) throw new Error(`Cannot forward dispatcher: ${key(sig)}`)
  const ret = some(lookup(inf, sig)).rettype
  const code = MIR(Def(`${F.name.path} (forwarder)`))
  const args = S.map((S, i) =>
    code.push(code.stmt(expr<IRValue>('cast', code.argument(S)), { type: T[i] })))
  code.return(code.push(code.stmt(new Invoke(F, args), { type: ret })))
  return [code, ret]
}

class Inferred implements Caching {
  readonly inf: Inference
  results: CacheMap<string, [MIR, Anno<Type>]>
  time = 0n

  constructor(defs: Definitions, lowered: Lowered, meths: MatchMethods, traced?: Traced, transforms?: Inferred) {
    this.inf = new Inference(defs, lowered, meths, traced, transforms)
    this.results = new CacheMap()
  }

  get size() { return this.results.size }
  iscached(k: string): boolean { return this.results.iscached(k) }

  _get(sig: Sig): [MIR, Anno<Type>] {
    const k = key(sig)
    if (this.iscached(k)) return this.results.get(k)!
    // Don't let inference dependencies leak
    const [_, deps] = trackdeps(() => {
      call(this.inf, new Stack(), sig)
      infer(this.inf)
    })
    if (deps.size !== 0) throw new Error('assertion')
    for (const r of this.inf.redirects.keys()) this.results.delete(r)
    for (const [k, fr] of this.inf.frames) {
      if (this.iscached(k)) continue
      this.results.set(k, [redirectCalls(this.inf, prune(unloop(fr.ir))), fr.rettype])
    }
    if (this.inf.redirects.has(k)) this.results.set(k, forward(this.inf, sig))
    return some(this.results.get(k))
  }

  get(sig: Sig): [MIR, Anno<Type>] {
    const [res, t] = withtime(() => this._get(sig))
    this.time += t
    return res
  }

  fingerprint(): Set<bigint> { return fingerprint(this.results) }

  reuse(ch: this): this {
    this.inf.reuse(ch.inf)
    this.results = ch.results.clone()
    return this
  }

  reset(deps: Set<bigint>) {
    for (const [x, d] of this.inf.deps) {
      const sub = Array.from(d).every(id => deps.has(id))
      if (!sub) remove(this.inf, x)
    }
    for (const k of this.results.keys())
      if (!this.inf.frames.has(k)) this.results.delete(k)
  }

  traitType(T: Type): Anno<Type> {
    return traitResult(this.get(traitSig(T))[1])
  }
}

function traitSig(T: Type): [Dispatch, Type, Type] {
  const f = tag('common.patterns/castTrait')
  return [new Dispatch(f), f, pack(tag('common.list/List'), T, Any)]
}

function traitResult(ret: Anno<Type>): Anno<Type> {
  if (ret === unreachable) return unreachable
  for (const option of disjuncts(ret))
    if (tag('common.core/Optional.Some').isEqual(tagOf(option))) return only(parts(option))
  return unreachable
}

const [withTraits, getTraits] = binding<(T: Type) => Anno<Type>>('traits')

function traitType(T: Type): Anno<Type> { return getTraits()(T) }
